### PR TITLE
Read Memos from file.

### PR DESCRIPTION
**Pull request by developer88**
While working on some project faced the file, that contains only pointers inside memo-field to external file. More info  [here](https://msdn.microsoft.com/en-us/library/aa975373%28v=vs.71%29.aspx). Found a solution [here](http://stackoverflow.com/questions/1947348/how-to-read-foxpro-memo-with-php) and adapted it for XBase.

### CODE REVIEW

**Comment by hisamu** [comment]
Hi @developer88,

You did some changes in the code that I can't accept as it is. Your commit implements e second parameter in the `__construct` method, before two other parameters that already existed, one of them exists since I first commited the code. If I accept it, it will probably break compatibility with several other users of the lib.

I believe we could implement this with other approach. I propose those changes:

- The `open` method should not be renamed. It's used in other places and it should stay as it were.
- The `$memoFileName` var should be set in a setter method, after initialization. A `setMemoFile` method could be created, and inside it we could call a new `open` method, only for the memo, and also instantiate the `$mfp` property, which won't exist before this method is called.
- The `$hasMemos` property is never used, it should be removed.
- The `getMemoFromFile` method does not belong in the Record. It would probably fits better in the Table class, as the Record object does not seek into the table file.
- You could create a Memo class which handles the file reading and the pointer seek. Instantiating it inside the Table via the new `setMemoFile` or `setMemo` method would give the Record objects access to it and isolate all the Memo related code, leaving the core Table with the same structure.

Do you believe you could look into this changes?
I could help if I had a file sample of both the db that uses the memo and the memo itself, but I don't know when I'd have the time.

Thanks for sharing this.

**Comment by developer88** [comment]
Thanks @hisamu .I will make changes soon.

**Comment by chetstriker** [comment]
In order for this to work, I had to alter the Record.php as follows:

    ```
public function getMemoFromFile($pointer) {
        $Value = null;
        if($this->table->mfp && $pointer[1] != 0) {  //addedd [1] to pointer
            // Getting block size
            fseek($this->table->mfp, 6);
            $Data = unpack("n", fread($this->table->mfp, 2));
            $Memo_BlockSize = $Data[1];

            
            fseek($this->table->mfp, $pointer[1] * $Memo_BlockSize); // was fseek($this->table->mfp, $pointer * $Memo_BlockSize);
            $Type = unpack("N", fread($this->table->mfp, 4));
            if($Type[1] == "1") {
                $Len = unpack("N", fread($this->table->mfp, 4));
                $Value = trim(fread($this->table->mfp, $Len[1]));
                if ($this->table->getConvertFrom()) {
                    $Value = iconv($this->table->getConvertFrom(), 'utf-8', $Value);
                }
            } else {
                // Pictures will not be shown
                $Value = "{BINARY_PICTURE}";
            }
        }
        return $Value;
    }
```

```
    public function getMemo($columnName)
    {
        $data = $this->forceGetString($columnName);
        if($data && strlen($data) == 2) {
            $pointer = unpack('s', $data);  // was $pointer = unpack('s', $data)[1];
            return $this->getMemoFromFile($pointer);
        } else {
            return $data;
        }
    }
```

And the Table.php file as:

```
 public function __construct($tableName, $memoFileName = null, $avaliableColumns = null, $convertFrom = null)
      {
          $this->tableName = $tableName;
          $this->memoFileName = $memoFileName;
          $this->avaliableColumns = $avaliableColumns;
          $this->convertFrom = $convertFrom;

        


          $this->open_foxpro_file($this->fp, $this->tableName, true);

          if($this->memoFileName!==null)
          {
              //only call this if a memo is attached
          $this->open_foxpro_file($this->mfp, $this->memoFileName);
          }

          
      }
```

**Comment by developer88** [comment]
hi @hisamu . it took me ages to prepare fixes. This is just a draft. I will squash this later. Do you agree with my changes?

File: src/XBase/Record.php [hunk 1]
```diff
@@ -188,7 +188,13 @@ public function getBoolean($columnName)
 
     public function getMemo($columnName)
     {
-        return $this->forceGetString($columnName);
+        $data = $this->forceGetString($columnName);
```
Comment (developer88) [inline src/XBase/Record.php]: should i move this to Memo class too?

**Comment by developer88** [comment]
anyone here?)

File: src/XBase/Memo.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        return $this->fp != false;
+    }
+
+    public function get($pointer) {
+        $Value = null;
+        if($this->fp && $pointer != 0) {
+            // Getting block size
+            fseek($this->fp, 6);
+            $Data = unpack("n", fread($this->fp, 2));
+            $Memo_BlockSize = $Data[1];
```
Comment (developer88) [inline src/XBase/Memo.php]: Good idea. Give me some time please 😄

**Comment by developer88** [comment]
@hisamu i did some testing, bugfixed some stuff and did renaming for variables.

**Pull request merged**